pageserver/src/tenant/remote_timeline_client.rs

```diff
@@ -768,6 +819,89 @@ impl RemoteTimelineClient {
         Ok(())
     }
 
+    pub(crate) async fn delete_all(self: &Arc<Self>) -> anyhow::Result<()> {
+        let mut receiver = {
+            let mut locked = self.upload_queue.lock().unwrap();
+            let stopped = match &mut *locked {
+                UploadQueue::Uninitialized => {
+                    anyhow::bail!("upload queue is not stopped but initialized")
+                }
+                UploadQueue::Initialized(_) => {
+                    anyhow::bail!("upload queue is not stopped but initialized")
+                }
+                UploadQueue::Stopped(stopped) => stopped,
+            };
+
+            if !matches!(stopped.deleted_at, SetDeletedFlagProgress::Successful(_)) {
+                anyhow::bail!("deleted_at is not set")
+            }
+
+            debug_assert!(stopped
+                .upload_queue_for_deletion
+                .inprogress_tasks
+                .is_empty());
+            debug_assert!(stopped
+                .upload_queue_for_deletion
+                .queued_operations
+                .is_empty());
+
+            // schedule the actual deletions
+            for name in stopped.upload_queue_for_deletion.latest_files.keys() {
+                let op = UploadOp::Delete(Delete {
+                    file_kind: RemoteOpFileKind::Layer,
+                    layer_file_name: name.clone(),
+                    scheduled_from_timeline_delete: true,
+                });
+                self.calls_unfinished_metric_begin(&op);
+                stopped
+                    .upload_queue_for_deletion
+                    .queued_operations
+                    .push_back(op);
+                info!("scheduled layer file deletion {}", name.file_name());
```

Comment:
to allocate memory once consider using queued_operations.reserve(stopped.upload_queue_for_deletion.latest_files.keys().size())

Reply:
Done